### PR TITLE
api: add /storage_service/describe_ring/{keyspace}/{table}

### PR DESCRIPTION
**Pull request by tchaikov**
this API is added to support the tablet keyspace, which is a
per-table replica strategy, with which each table is composed
of multiple tablets, each of which is replicated on its own
replicas. so, in order to schedule repair, scylla manager needs
to understand the topology of keyspace on a per-table basis instead
of on a per-keyspace basis.

in this change, the new RESTful API of
`/storage_service/describe_ring/{keyspace}/{table}` enables us to
retrieve a list of token range and its corresponding endpoints with
given table.

Refs https://github.com/scylladb/scylladb/issues/16509

### CODE REVIEW

**Comment by tchaikov** [comment]
it's still missing a test for this API. will add it shortly.

File: service/storage_service.cc [hunk 1]
```diff
@@ -4409,6 +4409,11 @@ storage_service::describe_ring(const sstring& keyspace, bool include_only_local_
     return locator::describe_ring(_db.local(), _gossiper, keyspace, include_only_local_dc);
 }
 
+future<std::vector<storage_service::token_range_endpoints>>
+storage_service::describe_ring_for_table(std::string_view ks_name, std::string_view cf_name) const {
+    return locator::describe_ring_for_table(_db.local(), _gossiper, ks_name, cf_name);
+}
```
Comment (denesb) [inline service/storage_service.cc:4415]: Looks like this hunk doesn't belong here.

File: locator/util.cc [hunk 2]
```diff
@@ -148,4 +147,11 @@ describe_ring(const replica::database& db, const gms::gossiper& gossiper, const
     co_return ranges;
 }
 
+future<std::vector<dht::token_range_endpoints>>
+describe_ring(const replica::database& db, const gms::gossiper& gossiper, const sstring& keyspace, bool include_only_local_dc) {
+    //Token.TokenFactory tf = getPartitioner().getTokenFactory();
```
Comment (denesb) [inline locator/util.cc:152]: I think we can drop this comment. It probably remained here from the time when we rewrote C* Java code to C++.

**Comment by mykaul** [comment]
CC @Michal-Leszczynski , @karol-kokoszka

**Review by denesb — COMMENTED** [review]
LGTM, some small comments.

Now we just need to implement `scylla-nodetool describering` with support for this endpoint (in addition to the legacy per-keyspace one).

File: locator/util.cc (hunk 2 above)
Comment (tchaikov) [reply locator/util.cc]: added a dedicated commit to drop this comment.

File: service/storage_service.cc (hunk 1 above)
Comment (tchaikov) [reply service/storage_service.cc:4415]: ahh, exactly. moved it to the place where it belongs to.

**Comment by tchaikov** [comment]
> LGTM, some small comments.
> 
> Now we just need to implement `scylla-nodetool describering` with support for this endpoint (in addition to the legacy per-keyspace one).

thank you for the suggestion. will do early tomorrow.

**Comment by denesb** [comment]
> thank you for the suggestion. will do early tomorrow.

This can be a follow-up, this PR can go in as-is.

**Comment by tchaikov** [comment]
v2:

- added a commit to drop stale comment of the original java implmenetation.
- moved the storage_service change to the 2nd commit

File: service/storage_service.cc (hunk 1 above)
Comment (denesb) [reply service/storage_service.cc:4415]: I still see a change here, even though `locator/util.cc` the method is still called `describe_ring()`.

Comment (tchaikov) [reply service/storage_service.cc:4415]: i am sorry, should be good now.

**Comment by denesb** [comment]
@asias reaised that we should reject calls to the keyspace-specific API (which works for vnodes), if the keyspace has tablets enabled. Similarly, for the new API introduced in this PR, we should reject non-tablet keyspaces.

**Comment by asias** [comment]
@tchaikov I took a look at your patch, I think we can keep the existing api and just add a table parameter to make it work for both vnode and tablet table. Another thing is that we can use the tmap.for_each_tablet and tablet_info.replicas to get the token ranges directly without all the legacy tokens and ranges calculation with get_range_to_address_map and friends.  Like mentioned by Botond, we need some checks to avoid incorrect usage by user.  So, I send my version here https://github.com/scylladb/scylladb/pull/17118 too for your consideration.

**Comment by tchaikov** [comment]
@asias thank you for the reviews, and for sending the PR, i am closing this PR.  just to avoid the overhead of repeating works, are you taking #16509 ?

**Comment by asias** [comment]
> @asias thank you for the reviews, and for sending the PR, i am closing this one. just to avoid the overhead of repeating works, are you taking #16509 ?

Yes, next, I plan to add the range parameter support for the tablet repair. I will continue to add all the necessary changes needed in the scylla side to make scylla manger works with tablet table repair.

**Comment by tchaikov** [comment]
> > @asias thank you for the reviews, and for sending the PR, i am closing this one. just to avoid the overhead of repeating works, are you taking #16509 ?
> 
> Yes, next, I plan to add the range parameter support for the tablet repair. I will continue to add all the necessary changes needed in the scylla side to make scylla manger works with tablet table repair.

okay, i am assigning #16509 to you then. please let me know if i can help.

**Pull request closed without merging**